rd_to_qmd_to_md 2

rempsyc opened the pull request:
Continuation from #53 so I can bring changes directly.

---
> Examples should be executed and rendered, but I have not tried examplesIf tags, so this should be tested.
- [x] Examples executed 
- [x] Examples rendered
- [x] Use proper function/section names 
- [x] Test examplesIf tags

Closes #33 (I think)

rempsyc commented:
Ok, so it took forever to render `marginaleffects` but it seems to have executed the examples (as shown in the console). 

![reference rendered](https://github.com/etiennebacher/altdoc/assets/13123390/9c58ca8a-4959-4714-bdf5-1fc2209e8d27)

However, the site doesn't show the Reference section at all it seems at this point, I probably need to make some other adjustments:

![reference](https://github.com/etiennebacher/altdoc/assets/13123390/6782f081-b63d-4f9e-aa23-6a83f4eb47d3)

vincentarelbundock commented:
Looks like maybe you rendered all the internal functions as well? `set_coef()` is not supposed to be documented for users...

rempsyc commented:
On the face of it, it seems like `.rd2md()` and `.rd_to_qmd()` should have similar functionality. Therefore I thought of doing a simple adaptation, like,

```r
all_rd_as_md <- lapply(files, function(x){
    if (quarto) {
      .rd_to_qmd(x)     
    } else {
      .rd2md(x)
    }
  })
```

But this does not work, because `.rd_to_qmd()` also requires a `target_dir` argument, because it converts each `Rd` file to an `md` equivalent, so we end up with as many `md` files as we had `Rd`. `.rd2md()` differs in two aspects. First, it combines all the `Rd` files in a single file. Second, instead of saving to disk, it saves the output to an object. Therefore, a bit of adaptation will be necessary.

rempsyc commented:
Ok with the latest changes, it seems the examples render and are also executed :D

![examples rendered](https://github.com/etiennebacher/altdoc/assets/13123390/6be3fec1-f923-4ff4-aa23-21be246c280c)

Only thing is that so far the correct function names do not appear. This should be corrected next.

rempsyc commented:
Now supports proper function/section names, but `## Not run:` is run:

![proper function names](https://github.com/etiennebacher/altdoc/assets/13123390/ef325e59-6252-4354-b643-b51deea5d506)

rempsyc commented:
In `marginaleffects`, there is a single use of `@examplesIf`, for `slopes.R`, but it is empty and followed by a regular `@examples` section. I also can't seem to redocument this package (experiencing an error, probably because it uses a book format) to test `@examplesIf` with `marginaleffects` directly.

It is also difficult to test in other packages, since most don’t load their own package in examples (and thus produce an error when rendering). I also could not properly redocument `altdoc` after adding an `@examplesIf` tag to `altdoc` examples themselves (seems my changes are ignored somehow in the rendered `Rd`). The workaround I used was modifying the `Rd` files directly by hand, using:

```r
\examples{
\dontshow{if (FALSE) (\{ # examplesIf}
1+1
\dontshow{\}) # examplesIf}
}
```
Using this, it seems that `@examplesIf` are rendered even when the condition is not met. Vincent do you want me to fix this before merging this PR? It is not technically a downgrade since it is regarding a new feature. Perhaps we could address it in a separate PR?

rempsyc commented:
> If there are no backward changes and the improvements are unambiguous, we could aim to use this as the main mechanism to convert man pages to appropriate markdown in all formats.

Does that mean that if we set `quarto = TRUE` and successfully merge this PR as default, we can close #44?

vincentarelbundock commented:
Yep, I think we can close those issues.

You can open a separate issue for example if

Maybe you can call pkgload::load_all() automatically before rendering the docs. That way, even if package authors don't call library() explicitly, the examples should run.

vincentarelbundock commented:
Let me know when you're ready for review.

rempsyc commented:
For `@examplesIf`, I opened #57 but then closed it because Etienne reminded me we already have #45.

For `pkgload::load_all()`, I tried adding it before running `.rd_to_qmd()` and `.qmd_to_md()`, and also within `.qmd_to_md()`, but as expected, that didn't work, because quarto renders the vignettes in a new environment. I also don't see an `envir` argument in `quarto::quarto_render()` like we have in `rmarkdown::render()`, so not sure this is an easy fix.

So in light of these, I think this is ready for review.

vincentarelbundock commented:
I added one comment with a feature request, but aside from that things look really great!

rempsyc commented:
@vincentarelbundock it seems your comment has not been "submitted" so I can't see it right now (happens to me all the time)

vincentarelbundock commented:
My only review is that there needs to be an option to only convert exported functions. Currently, a lot of the man/ pages you convert are documented internal functions, which takes forever and will not be displayed on the website

rempsyc commented on R/reference.R:
```diff
@@ -18,9 +18,17 @@
   files <- list.files("man", full.names = TRUE)
```
When I render `marginaleffects`, the two top functions are: `comparisons` (which is exported; `@export`), and `complete_levels` (which is not exported; `@keywords internal`). I see. Looking at `Rd` files in `man`, I don't see an export tag in `comparisons.Rd`. The only tag that seems preserved is `\keyword{internal}`. Do you want me to keep exported files using a negative logic (exclude those with `\keyword{internal}`), knowing that not everybody uses this? Or do you have a better idea? One possibility is to parse the R files looking for the export tag, but should there not be an easier solution?

rempsyc replied:
When I render `marginaleffects`, the two top functions are: `comparisons` (which is exported; `@export`), and `complete_levels` (which is not exported; `@keywords internal`). I see. Looking at `Rd` files in `man`, I don't see an export tag in `comparisons.Rd`. The only tag that seems preserved is `\keyword{internal}`. Do you want me to keep exported files using a negative logic (exclude those with `\keyword{internal}`), knowing that not everybody uses this? Or do you have a better idea? One possibility is to parse the R files looking for the export tag, but should there not be an easier solution?

vincentarelbundock replied:
No we don't want our solution to be roxygen2 specific. We need this to work with other packages too, and parsing is error prone.

There are many options like looking at NAMESPACE or this function: https://stackoverflow.com/a/9675792/342331

rempsyc replied:
Ah, great, thanks!

rempsyc replied:
Ok, I got something running, but the problem is that for example, for `altdoc`, the file `init.Rd` contains all three `use_*` functions. So even though I can get the list of exported functions through `getNamespaceExports(pkg)`, they don't match the `.Rd` file names. So it seems like I would still need to somehow select all `.Rd` files, then sub-select by parsing each `.Rd` file to see if they include the exported function names from `getNamespaceExports(pkg)`...

vincentarelbundock replied:
Sure, just intersect() them.

rempsyc replied:
I don't think `intersect()` would work because that is like an inner joint: it only keeps elements present in both lists. But in the case of the `init.Rd` file, it doesn't match with any of the exported function names. Right now I'm trying some regex within the `.Rd` with `\name{`...

rempsyc replied:
Ok, not pretty but the following seems to work...

```r
files <- list.files("man", pattern = ".Rd", full.names = TRUE)
pkg <- basename(getwd())
exported <- getNamespaceExports(pkg)

which.files <- lapply(files, function(x) {
    y <- readLines(x)
    y <- grep("\\name{", y, fixed = TRUE, value = TRUE)
    y <- gsub("\\name{", "", y, fixed = TRUE)
    y <- gsub("}", "", y, fixed = TRUE)
    y %in% exported
})

files <- files[unlist(which.files)]
```

![only exported](https://github.com/etiennebacher/altdoc/assets/13123390/2e142207-658d-4d60-ac49-92517d0448e1)

vincentarelbundock replied:
Hmm, that's weird. I didn't expect all these `get_()` and `expect_()` functions to be exported. They are not user-facing. I guess as long as there's a way to manually set them, it's fine like that. Let me know when you'd like me to review.

rempsyc replied:
But in the file `get_coef.R`, you have

```r
#' @export
get_coef <- function (model, ...)
```

And so it is exported:

``` r
getNamespaceExports("marginaleffects")[32]
#> [1] "get_coef"
```

<sup>Created on 2023-11-05 with [reprex v2.0.2](https://reprex.tidyverse.org)</sup>

After the last push I got an error in tests, so I'll try fixing that and then it should be ready for review. I will let you know.

vincentarelbundock replied:
Yeah, you're right. I forgot about that. I'll still be able to specify manually, right?

rempsyc replied:
`use_docute()` and family have a `custom_reference` argument: 
```
#' @param custom_reference Path to the file that will be sourced to generate the
#' "Reference" section.
```
So yes :) I think if you provide your custom Reference file, it should work!

rempsyc commented:
The new error is:

```
── Error ('test-vignettes.R:23:3'): .transform_vignettes works on basic vignette ──
<packageNotFoundError/error/condition>
Error in `loadNamespace(name)`: there is no package called 'testpkg19fa2625e519'
```

This seems to be because we added a step to load the package in references, and now that we changed the default to `quarto = TRUE`, reference examples try to be executed. Although the examples use `if (interactive()) {}`, the `loadNamespace()` call happens before that. In tests we use tests packages, and because those are not installed, they are not able to be loaded. One solution would be to set quarto = FALSE for these tests.

rempsyc commented:
Ok even setting `quarto = FALSE`, I get the same error now. The error comes from `getNamespaceExports(pkg)` if the said package is not installed. Current workaround is to parse the NAMESPACE instead:

```r
exported <- readLines("NAMESPACE")
exported <- grep("^export\\(", exported, value = TRUE)
exported <- gsub("export\\((.*)\\)", "\\1", exported)
```

This is ready to merge, I think.

vincentarelbundock commented:
Looks great, thanks!

etiennebacher commented:
Thanks, next time can you use squash and merge to avoid cluttering the git history? :)

etiennebacher commented:
Ah yes actually I can disable merge commits and rebase, so that only squash and merge are possible
